src/main/java/com/example/springbootbookshop/service/impl/BookServiceImpl.java

```diff
@@ -47,8 +57,27 @@ public void deleteById(Long id) {
     public BookDto update(CreateBookRequestDto bookRequestDto, Long id) {
         Book bookToUpdate = bookRepository.findById(id)
                 .orElseThrow(() -> new EntityNotFoundException("Can`t find book with id:" + id));
+        Set<Category> categories = getCategories(bookRequestDto);
         bookMapper.updateBook(bookRequestDto, bookToUpdate);
-        bookRepository.save(bookToUpdate);
-        return bookMapper.toDto(bookToUpdate);
+        bookToUpdate.setCategories(categories);
+        return bookMapper.toDto(bookRepository.save(bookToUpdate));
+    }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id) {
+        if (categoryRepository.existsById(id)) {
+            return bookRepository.findAllByCategoryId(id).stream()
+                    .map(bookMapper::toDtoWithoutCategories)
+                    .toList();
+        }
+        throw new EntityNotFoundException("Category with id " + id
+                + " doesn`t exist");
+    }
+
+    private Set<Category> getCategories(CreateBookRequestDto bookDto) {
+        return bookDto.categoryIds().stream()
+                .map(id -> categoryRepository.findById(id).orElse(null))
```

Comment:
too may requests to DB

either use findAllByIds or getReferenceById (preferable)

Reply:
I need help, if you use getreference, there is a problem with the initialization exception and I don't know how to fix it, I'm a little confused